src/cache_padded.rs

```diff
@@ -58,63 +42,102 @@ zeros_valid!(i8 i16 i32 i64 isize);
 unsafe impl ZerosValid for ::std::sync::atomic::AtomicUsize {}
 unsafe impl<T> ZerosValid for ::std::sync::atomic::AtomicPtr<T> {}
 
-impl<T: ZerosValid> CachePadded<T> {
-    /// A const fn equivalent to mem::zeroed().
-    #[cfg(not(feature = "nightly"))]
-    pub fn zeroed() -> CachePadded<T> {
-        CachePadded {
-            data: UnsafeCell::new(([0; CACHE_LINE])),
-            _marker: ([], marker::PhantomData),
-        }
-    }
+/// Assert that the size and alignment of `T` are consistent with `CachePadded<T>`.
+fn assert_valid<T>() {
+    assert!(mem::size_of::<T>() <= mem::size_of::<CachePadded<T>>());
+    assert!(mem::align_of::<T>() <= mem::align_of::<CachePadded<T>>());
+}
+
+/// Pad `T` to the length of a cacheline.
+///
+/// Sometimes concurrent programming requires a piece of data to be padded out to the size of a
+/// cacheline to avoid "false sharing": cachelines being invalidated due to unrelated concurrent
+/// activity. Use the `CachePadded` type when you want to *avoid* cache locality.
+///
+/// # Warning
+///
+/// - The wrapped data is never dropped; move out using `ptr::read` if you need to run dtors.
+/// - Do not rely on this actually being padded to the cache line for the correctness of your
+///   program. Certain compiler options or hardware might actually cause it to not be aligned.
+pub struct CachePadded<T> {
+    /// The data.
+    ///
+    /// This is untyped on purpose in order to force a certain memory representation.
+    data: UnsafeCell<[u8; CACHE_LINE]>,
+    /// The memory padding.
+    _pad: Pad,
+    /// The typed marker.
+    _marker: marker::PhantomData<T>,
+}
 
-    /// A const fn equivalent to mem::zeroed().
+impl<T: ZerosValid> CachePadded<T> {
+    /// Create a zeroed cache padded value.
     #[cfg(feature = "nightly")]
     pub const fn zeroed() -> CachePadded<T> {
```

Comment:
In that case how about adding `assert_valid` to `default` and `const fn zeroed` to address the `ZerosValid` size assertion issue?